Add support for ewasm evmc vm.

aarlt opened the pull request:
Closes #8016.

Replaces #9057.

This PR enables the possibility to run tests against `ewasm` evmc vms.
 
In contrast to #9057, where n-evmc vms could be defined, this PR is more simple. It enables the definition of up-to two different evmc libaries. To execute the tests, at least one `evm1` vm must be defined (either through the `--evmonepath` parameter, or with `--vm`), where the defintion of an `ewasm` vm is optional. If an `ewasm` evmc library was defined (via `--vm`), tests will also run against that ewasm evmc vm.

- [X] Basic structure
- [X] SemanticTest: add support for `compileToEwasm`
- [x] Basic CircleCI integration to also run Semantic Tests via ewasm vm (moved to #9212)

aarlt committed:
Add support for ewasm evmc vm.

chriseth commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -84,157 +84,164 @@ SemanticTest::SemanticTest(string const& _filename, langutil::EVMVersion _evmVer
+							soltestAssert(
+								m_allowNonExistingFunctions
+									|| m_compiler.methodIdentifiers(m_compiler.lastContractName())
+										   .isMember(test.call().signature),
```
This is quite some indentation depth. Can we somehow fix this by introducing helper functions?

axic commented:
@aarlt we are working on landing https://github.com/ewasm/hera/pull/554. It will be released as Hera 0.3.0.

axic commented:
@aarlt done, available here: https://github.com/ewasm/hera/releases/tag/v0.3.0

aarlt commented on libsolidity/codegen/ir/IRGenerator.cpp:
```diff
@@ -583,7 +583,9 @@ string IRGenerator::dispatchRoutine(ContractDefinition const& _contract)
 		if iszero(lt(calldatasize(), 4))
 		{
-			let selector := <shr224>(calldataload(0))
+			mstore(0, 0)
```
This is just a copy from #7597.

aarlt replied:
I also added a comment to #7597: would it make sense to generate different code here, depending on evm / ewasm?

chriseth commented on .circleci/install_hera.sh:
```diff
@@ -0,0 +1,40 @@
+# ------------------------------------------------------------------------------
+
+apt-get update
+apt-get upgrade -y
```
Is this really needed? It makes the image less predictable.

aarlt replied:
yeah, it is needed, at least for now.. it looks like that there is a problem with the current gcc that is installed within the docker images. The installed gcc had problems compiling hera.. with an error like that one:
```
/usr/lib/gcc/x86_64-linux-gnu/9/include-fixed/bits/statx.h:38:25: error: missing binary operator before token "("
   38 | #if __glibc_has_include ("__linux__/stat.h")
      |  
```
After upgrading gcc the compilation was successful. We would need to recreate the docker images and update the packages to the latest version.

chriseth commented on test/cmdlineTests/standard_eWasm_requested/output.json.orig:
```diff
@@ -0,0 +1,354 @@
+        (local.set $hi (i64.shl (call $endian_swap_16 (local.get $x)) (i64.const 16)))
+        (local.set $y (i64.or (local.get $hi) (call $endian_swap_16 (i64.shr_u (local.get $x) (i64.const 16)))))
+
+<<<<<<< HEAD
```
conflict marker

aarlt replied:
ah, looks like that I accidentally checked in that file.

chriseth commented on test/cmdlineTests/yul_string_format_ascii_long/output.json:
```diff
@@ -28,7 +28,9 @@ object \"C_10\" {
 
             if iszero(lt(calldatasize(), 4))
             {
-                let selector := shift_right_224_unsigned(calldataload(0))
```
Can you check what the gas cost difference here is?

aarlt replied:
I did some measurements last week. See https://github.com/ethereum/solidity/pull/7597#discussion_r439523065

aarlt replied:
@chriseth @axic what should we do here now? should we include commit c8d95d601ee540e5d33e05b7fe718adb44a20dea? or will hera be changed in a way that the old code will just work?

aarlt commented:
Some test-cases seem to hang inside hera (e.g. `nested_for_loop_multiple_local_vars`). It looks like that this is mainly related to tests using loops that generate infinity loops. Right now metering is not enabled for the ewasm tests. I think some tests hang in hera because of not enabled metering and wrongly generated code leading to infinity loops.

chriseth commented on test/Common.cpp:
```diff
@@ -92,6 +92,7 @@ CommonOptions::CommonOptions(std::string _caption):
 		("evm-version", po::value(&evmVersionString), "which evm version to use")
 		("testpath", po::value<fs::path>(&this->testPath)->default_value(solidity::test::testPath()), "path to test files")
 		("evmonepath", po::value<fs::path>(&evmonePath)->default_value(EVMOneEnvOrDefaultPath()), "path to evmone library")
+		("vm", po::value<std::vector<fs::path>>(&vmPaths), "path to evmc library. can be defined multiple times.")
```
```suggestion
		("vm", po::value<std::vector<fs::path>>(&vmPaths), "path to evmc library, can be supplied multiple times.")
```

chriseth commented on test/Common.h:
```diff
@@ -44,6 +44,7 @@ struct ConfigException : public util::Exception {};
 struct CommonOptions: boost::noncopyable
 {
 	boost::filesystem::path evmonePath;
```
Does it matter which one is the evmone path?

aarlt replied:
I left `evmonePath` here to ensure the old behaviour of defining this via `--evmonepath`. `evmonePath` may also be defined automatically, if an evmone library was found in default paths.

Internally `vmPaths` is always used. If no shared library is defined via `--vm`, `evmonePath` will always be added to the `vmPaths`. See [`CommonOptions::parse(..)`](https://github.com/ethereum/solidity/blob/583bbbb5a463dfe1e9d8504883b9c53a9a9977b0/test/Common.cpp#L144).

But this also means, that in the situation where just one library was defined with`--vm`, the evmone library need to be defined via `--vm` as well.

chriseth replied:
I think it would be better to do it properly. This is a test runtime, we can make breaking changes any time.

aarlt replied:
I removed now `--evmonepath`. Multple evmc vms can now be only defined with `--vm`. If no vms where defined with `--vm`, it will search for an `evmone` library automatically. if additionally `--ewasm` was defined, it will automatically search for `hera` too.

chriseth commented on test/ExecutionFramework.cpp:
```diff
@@ -39,27 +41,54 @@ using namespace solidity::util;
 	if (solidity::test::CommonOptions::get().optimize)
 		m_optimiserSettings = solidity::frontend::OptimiserSettings::standard();
 
+	for (auto const& path : _vmPaths)
```
```suggestion
	for (auto const& path: _vmPaths)
```

chriseth commented on test/ExecutionFramework.h:
```diff
@@ -275,7 +277,13 @@ class ExecutionFramework
 	solidity::frontend::OptimiserSettings m_optimiserSettings = solidity::frontend::OptimiserSettings::minimal();
 	bool m_showMessages = false;
+
+	std::shared_ptr<EVMHost> m_evmcHost;
```
Why are there three?

aarlt replied:
`m_evmcHost` is the "current" evmc host that is used for all interactions. The two others are an `evm1` evmc host `m_evmHost` and the `ewasm` evmc host `m_ewasmHost`. The logic is basically to set `m_evmcHost` to `m_evmHost` or `m_ewasmHost`, depending on what kind of evmc vm should be used and what vm's where defined. 

The switch is mainly done in `ExecutionFramework::reset(bool useEwasm)`, if `useEwasm` is true and there was a evmc vm defined supporting ewasm, `m_evmcHost` will be set to `m_ewasmHost`. If `useEwasm` is false, `m_evmcHost` is set to `m_evmHost`.

chriseth replied:
Ah, ok, I see. Can you make it a plain (non-owning) pointer then?

chriseth commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -84,157 +84,164 @@ SemanticTest::SemanticTest(string const& _filename, langutil::EVMVersion _evmVer
 	for (bool compileViaYul: set<bool>{!m_runWithoutYul, m_runWithYul || m_enforceViaYul})
-	{
-		try
+		for (bool compileToEwasm: compileViaYul && supportsEwasm() ? set<bool>{false, true} : set<bool>{false})
```
Is it possible to keep this at a single indentation?

chriseth commented on test/EVMHost.cpp:
```diff
@@ -38,17 +38,17 @@ using namespace evmc::literals;
 {
-	static evmc::VM theVM;
-	if (!theVM && !_path.empty())
+	static std::map<string, std::unique_ptr<evmc::VM>> vms;
```
Why unique_ptr instead of evmc::VM directly?

chriseth commented on test/EVMHost.cpp:
```diff
@@ -58,7 +58,45 @@ evmc::VM& EVMHost::getVM(string const& _path)
+	if (vms.count(_path) > 0)
+		return *vms[_path];
+
+	static evmc::VM NullVM(nullptr);
```
Can you move this up to the other static vms?

chriseth commented on test/EVMHost.cpp:
```diff
@@ -58,7 +58,45 @@ evmc::VM& EVMHost::getVM(string const& _path)
+		{
+			if (evmVmFound)
+			{
+				cerr << "Multiple evm1 evmc vms where defined. Please only define one evm1 evmc vm." << endl;
```
```suggestion
				cerr << "Multiple evm1 evmc vms defined. Please only define one evm1 evmc vm." << endl;
```

chriseth commented on test/EVMHost.h:
```diff
@@ -44,6 +46,8 @@ class EVMHost: public evmc::MockedHost
 	/// afterwards.
 	static evmc::VM& getVM(std::string const& _path = {});
 
+	static bool checkVmPaths(std::vector<boost::filesystem::path> const& _vmPaths);
```
Please document that it returns true if and only if an EVM VM was found (ignoring ewasm failures) and throws if multiple VMs of the same type are provided.

chriseth commented on test/EVMHost.h:
```diff
@@ -44,6 +46,8 @@ class EVMHost: public evmc::MockedHost
 	/// afterwards.
```
The documentation is not correct anymore.

chriseth commented on test/ExecutionFramework.cpp:
```diff
@@ -39,27 +41,54 @@ using namespace solidity::util;
+bool ExecutionFramework::supportsEwasm() const
+{
+	if (m_ewasmHost)
+		return m_ewasmHost->hasCapability(EVMC_CAPABILITY_EWASM);
```
Why wouldn't it?

chriseth commented on test/ExecutionFramework.cpp:
```diff
@@ -39,27 +41,54 @@ using namespace solidity::util;
+	return false;
+}
+
+void ExecutionFramework::reset(bool useEwasm)
```
I think this is a weird way to do it. Can you add a function that switches the VM and keep this reset function just resetting? (the switching function could of course call the reset function)

chriseth commented on test/ExecutionFramework.h:
```diff
@@ -57,7 +57,7 @@ class ExecutionFramework
 public:
 	ExecutionFramework();
-	explicit ExecutionFramework(langutil::EVMVersion _evmVersion);
+	explicit ExecutionFramework(langutil::EVMVersion _evmVersion, std::vector<boost::filesystem::path> const& _vmPaths);
```
```suggestion
	ExecutionFramework(langutil::EVMVersion _evmVersion, std::vector<boost::filesystem::path> const& _vmPaths);
```

chriseth commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -39,34 +39,46 @@ using namespace boost::unit_test;
+	else if (compileToEwasm == "false")
+		runWithEwasm = false;
+	else
+		BOOST_THROW_EXCEPTION(runtime_error("Invalid compileToEwasm value: " + compileToEwasm + "."));
```
Please add a check that runWithYul is true if runWithEwasm is true.

chriseth commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -84,159 +96,170 @@ SemanticTest::SemanticTest(string const& _filename, langutil::EVMVersion _evmVer
-			reset();
-			bool success = true;
+		for (bool compileToEwasm: runWithEwasm ? set<bool>{false, true} : set<bool>{false})
+			try
```
Can you pull this out into its own function?

chriseth commented:
What is the status here?

axic commented on libyul/backends/wasm/EVMToEwasmTranslator.cpp:
```diff
@@ -841,7 +841,13 @@ function callvalue() -> z1, z2, z3, z4 {
 	z1, z2, z3, z4 := mload_internal(0:i32)
 }
 function calldataload(x1, x2, x3, x4) -> z1, z2, z3, z4 {
```
I'd suggest to take this into its own PR and also fix codecopy, datacopy, extcodecopy, returndatacopy to truncate based on size.

chriseth replied:
I agree

aarlt replied:
I also agree, but at first I want to fix the polyfills here and extract those into a new PR later. It is more easy to test the changes here. Maybe I will also extract the semantic tests for ewasm to an other PR. But for now  I use this PR to ensure that the changes work.

aarlt commented:
> What is the status here?

@chriseth I think it's ready. I split out the CircleCI integration #9481 and currently trying to enable as much tests as possible while fixing some polyfills in #9508. From my point of view this PR #9143 and the CircleCI #9481 could get merged. All ewasm based tests are disabled in #9481 and #9143. Related tests get enabled in #9508. 

One thing: I added in a new CircleCI job `t_osx_test` in #9481 that will also run `soltest_all.sh` in osx. If #9508 is passing the tests now, it would be nice to merge #9481 and #9143. And then we can fully focus on getting #9508 ready.

chriseth commented on test/Common.cpp:
```diff
@@ -141,6 +142,9 @@ bool CommonOptions::parse(int argc, char const* const* argv)
 			throw std::runtime_error(errorMessage.str());
 		}
 
+	if (!evmonePath.empty() && vmPaths.empty())
```
Shouldn't this produce an error if `--vm` and `--evmonepath` is used together?

chriseth commented on test/EVMHost.cpp:
```diff
@@ -39,17 +39,18 @@ using namespace evmc::literals;
+				vms[_path] = evmc::VM(std::move(vm));
 			else
-				cerr << "VM loaded does not support EVM1" << endl;
+				cerr << "VM loaded does not support EVM1 or EWASM" << endl;
```
neither supports ... nor ...

chriseth commented on test/EVMHost.cpp:
```diff
@@ -59,7 +60,44 @@ evmc::VM& EVMHost::getVM(string const& _path)
+	return NullVM;
+}
+
+bool EVMHost::checkVmPaths(std::vector<boost::filesystem::path> const& _vmPaths)
```
```suggestion
bool EVMHost::checkVmPaths(vector<boost::filesystem::path> const& _vmPaths)
```

chriseth commented on test/EVMHost.cpp:
```diff
@@ -59,7 +60,44 @@ evmc::VM& EVMHost::getVM(string const& _path)
+		{
+			if (evmVmFound)
+			{
+				cerr << "Multiple evm1 evmc vms defined. Please only define one evm1 evmc vm." << endl;
```
Looks a bit weird to first print the error to stderr and then throw using an assertion and not specify any error.
Would it work to just throw a regular exception and print the error string at the place where the exception in caught?

aarlt replied:
i just remove the exception here. if something goes wrong the semantic tests will be disabled and the error is printed to stderr.

aarlt replied:
now i throw an exception, that will be caught in `test/boostTest.cpp` and `test/tools/isoltest.cpp`. In both cases the exception message will just get printed and the program will terminate.

chriseth commented on test/EVMHost.h:
```diff
@@ -40,11 +42,17 @@ class EVMHost: public evmc::MockedHost
 
+	/// Tries to load all defined shared libraries.
+	/// @param _vmPaths paths to multiple evmc shared libraries.
+	/// @throw Exception if multiple evm1 or multiple ewasm evmc vms where defined.
```
```suggestion
	/// @throw Exception if multiple evm1 or multiple ewasm evmc vms are defined.
```

chriseth replied:
or "loaded"?

chriseth commented on test/EVMHost.h:
```diff
@@ -40,11 +42,17 @@ class EVMHost: public evmc::MockedHost
+	/// Tries to load all defined shared libraries.
+	/// @param _vmPaths paths to multiple evmc shared libraries.
+	/// @throw Exception if multiple evm1 or multiple ewasm evmc vms where defined.
+	/// @returns true, if an evmc vm was defined supporting evm1,
+	///          false, otherwise, or if a defined evmc vm was not able to load.
```
```suggestion
	/// @returns true, if an evmc vm was supporting evm1 loaded properly.
```

chriseth commented on test/EVMHost.h:
```diff
@@ -71,6 +79,12 @@ class EVMHost: public evmc::MockedHost
 	static util::h256 convertFromEVMC(evmc::bytes32 const& _data);
 	static evmc::bytes32 convertToEVMC(util::h256 const& _data);
 
+	/// Checks if the VM has the given capability.
```
Which VM?

chriseth commented on test/EVMHost.h:
```diff
@@ -40,11 +42,17 @@ class EVMHost: public evmc::MockedHost
-	/// Tries to dynamically load libevmone. @returns nullptr on failure.
-	/// The path has to be provided for the first successful run and will be ignored
-	/// afterwards.
+	/// Tries to dynamically load an evmc vm supporting evm1 or ewasm.
```
Caches the loaded VM.

chriseth commented on test/EVMHost.h:
```diff
@@ -40,11 +42,17 @@ class EVMHost: public evmc::MockedHost
-	/// afterwards.
+	/// Tries to dynamically load an evmc vm supporting evm1 or ewasm.
+	/// @returns vmc::VM(nullptr) on failure.
 	static evmc::VM& getVM(std::string const& _path = {});
```
Is the default value still useful?

chriseth commented on test/ExecutionFramework.cpp:
```diff
@@ -40,27 +42,51 @@ using namespace solidity::util;
+	switchVm();
+}
+
+void ExecutionFramework::switchVm(bool _useEwasm)
```
Would it work to re-use the EVMC capabilities enum instead of the bool here?

aarlt replied:
good idea, using `evmc_capabilities` now

chriseth commented on test/ExecutionFramework.h:
```diff
@@ -279,7 +280,13 @@ class ExecutionFramework
+	EVMHost* m_evmcHost;
+
 	std::shared_ptr<EVMHost> m_evmHost;
+	std::shared_ptr<EVMHost> m_ewasmHost;
```
We call `reset` all the time on this - does it make sense to keep both around?

aarlt replied:
I was not sure where you exactly referring to. We have two real evm host instances, one for `evm` and one for `ewasm`. And one where `reset` is called `m_evmcHost`. This is just a pointer that may point either to the `evm` instance or to the `ewasm` instance.

chriseth replied:
I think my question was whether we should not have just the single member `unique_ptr<EVMHost> m_evmcHost` and recreate it whenever we would call reset.

aarlt replied:
implemented the recreation in`ExecutionFramework::setEvm(..)`.

chriseth commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -88,19 +103,29 @@ SemanticTest::SemanticTest(string const& _filename, langutil::EVMVersion _evmVer
 
 TestCase::TestResult SemanticTest::run(ostream& _stream, string const& _linePrefix, bool _formatted)
 {
+	set<pair<bool, bool>> viaYul;
```
This is getting really complicated. Can we extract the actual test run to a function that gets booleans about what it should compile? Then we maybe don't even have to run this in a weird loop over booleans but can just call this function multiple times in several lines?

aarlt replied:
Yes, I totally agree. I introduced a private method `runTest`, that is basically exactly the same as the original `run` method.  The current `run` method may now call `runTest` multiple times, it just depends whether `m_runWithoutYul`, `m_runWithYul || m_enforceViaYul` or `m_runWithEwasm` was set. I think it is much easier to read now.

aarlt commented:
> > What is the status here?
> 
> @chriseth I think it's ready. I split out the CircleCI integration #9481 and currently trying to enable as much tests as possible while fixing some polyfills in #9508. From my point of view this PR #9143 and the CircleCI #9481 could get merged. All ewasm based tests are disabled in #9481 and #9143. Related tests get enabled in #9508.
> 
> One thing: I added in a new CircleCI job `t_osx_test` in #9481 that will also run `soltest_all.sh` in osx. If #9508 is passing the tests now, it would be nice to merge #9481 and #9143. And then we can fully focus on getting #9508 ready.

I removed now the `t_osx_test` CircleCI job. Maybe we should think about how to extent the bytecode comparision tests with ewasm.

chriseth commented on test/Common.cpp:
```diff
@@ -57,9 +58,9 @@ boost::filesystem::path testPath()
 }
 
-std::string EVMOneEnvOrDefaultPath()
+std::string GetEnvOrDefaultPath(const char* env_name, const char* lib_name)
```
```suggestion
std::string envOrDefaultPath(char const* env_name, char const* lib_name)
```
But actually this should really use strings, even if you pass it on to functions that take C-strings.

chriseth commented on test/Common.h:
```diff
@@ -31,21 +33,27 @@ namespace solidity::test
-	boost::filesystem::path evmonePath;
+	std::vector<boost::filesystem::path> vmPaths;
 	boost::filesystem::path testPath;
+	bool enableEwasm = false;
```
Maybe the "enable" part is redundant?

aarlt replied:
Maybe `enableEwasm` is not the best name. I introduced `--ewasm` to the command-line options, if present `enableEwasm` is set to `true`. If enabled, a search for the hera evmc library will be done by using `envOrDefaultPath`. With this the ewasm tests can just be enable with `--ewasm` in our CI, because the libraries are there already in place. It is true that this is somehow redundant with specifying `--vm <path-to-hera>`, but I think it simplifies the testscripts.

chriseth replied:
I actually meant that this can be renamed to `bool ewasm`, but it's fine.

aarlt replied:
Ah, yes.. now I see what you meant :)

chriseth commented on test/EVMHost.cpp:
```diff
@@ -39,17 +39,18 @@ using namespace evmc::literals;
-	static evmc::VM theVM;
-	if (!theVM && !_path.empty())
+	static evmc::VM NullVM{nullptr};
+	static std::map<string, evmc::VM> vms;
```
Better use `map<string, unique_ptr<evmc::VM>>`, otherwise the returned reference might be invalidate if this function is called a second time.

chriseth commented on test/ExecutionFramework.cpp:
```diff
@@ -40,27 +42,47 @@ using namespace solidity::util;
 }
 
-ExecutionFramework::ExecutionFramework(langutil::EVMVersion _evmVersion):
+ExecutionFramework::ExecutionFramework(langutil::EVMVersion _evmVersion, std::vector<boost::filesystem::path> const& _vmPaths):
```
no `std::` in cpp files unless needed

chriseth commented on test/ExecutionFramework.cpp:
```diff
@@ -40,27 +42,47 @@ using namespace solidity::util;
+	{
+		evmc::VM& vm = EVMHost::getVM(path.string());
+		if (vm.has_capability(_cap))
+			m_evmcHost = std::unique_ptr<EVMHost, std::function<void(EVMHost*)>>(
```
Why the custom deleter?

chriseth replied:
Also maybe add a `break`?

aarlt replied:
In the header `EVMHost` is just forward declared. It is not possible to use the default-deleter on an incomplete type.

chriseth replied:
Ok sure, but why don't you properly declare it?

chriseth replied:
If it doesn't work without custom deleter then please use a shared_ptr instead of using a custom deleter.

aarlt replied:
Ah nice, it can be just properly declared. I was thinking that it would not work, because it was forward declared. Next time I will just try ;)

chriseth commented on test/ExecutionFramework.cpp:
```diff
@@ -40,27 +42,47 @@ using namespace solidity::util;
+	setEvm(evmc_capabilities::EVMC_CAPABILITY_EVM1);
+}
+
+void ExecutionFramework::setEvm(evmc_capabilities _cap)
```
`selectVM`?

aarlt replied:
yeah, better

chriseth commented on test/boostTest.cpp:
```diff
@@ -156,14 +155,27 @@ test_suite* init_unit_test_suite( int /*argc*/, char* /*argv*/[] )
 		cout << endl << "--- SKIPPING ALL SEMANTICS TESTS ---" << endl << endl;
+
+	cout << "Loaded EVMC VM's:" << endl;
+	for (auto &vmPath : solidity::test::CommonOptions::get().vmPaths)
```
```suggestion
	for (auto const& vmPath: solidity::test::CommonOptions::get().vmPaths)
```

chriseth commented on test/boostTest.cpp:
```diff
@@ -156,14 +155,27 @@ test_suite* init_unit_test_suite( int /*argc*/, char* /*argv*/[] )
+	for (auto &vmPath : solidity::test::CommonOptions::get().vmPaths)
+	{
+		evmc::VM& vm = solidity::test::EVMHost::getVM(vmPath.string());
+		cout << "- " << vm.name() << " v" << vm.version()  << " [caps=0x" << hex << vm.get_capabilities() << dec << ", path=" << vmPath.string() << "]" << endl;
```
Is this hex code readable?

aarlt replied:
Yes, it's either `0x1` for an evm1 vm or `0x2` for ewasm.

chriseth replied:
I would say this should not belong to the regular test output. `isoltest` should only print the names of the tests it runs or an error if something goes wrong.

aarlt replied:
Ok, I will remove that.

chriseth commented on test/boostTest.cpp:
```diff
@@ -156,14 +155,27 @@ test_suite* init_unit_test_suite( int /*argc*/, char* /*argv*/[] )
+	if (disableSemantics)
 		cout << endl << "--- SKIPPING ALL SEMANTICS TESTS ---" << endl << endl;
+
+	cout << "Loaded EVMC VM's:" << endl;
```
Does this get annoying over time?

aarlt replied:
I don't think so. I think it is good to see what evm's where actually loaded.

chriseth replied:
If you request ewasm and it cannot be loaded, the test should exit with an error - or do you really want to know the exact filename where it loaded the shared library from?

aarlt replied:
It was somehow useful while implementing.. but you're right, there is not much real value having this in general

chriseth commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -88,156 +103,180 @@ SemanticTest::SemanticTest(string const& _filename, langutil::EVMVersion _evmVer
-					_stream << endl;
+					soltestAssert(
+						m_allowNonExistingFunctions
+							|| m_compiler.methodIdentifiers(m_compiler.lastContractName())
```
`||` to the previous line.

chriseth commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -88,156 +103,180 @@ SemanticTest::SemanticTest(string const& _filename, langutil::EVMVersion _evmVer
+						"The function " + test.call().signature + " is not known to the compiler");
+
+					output = callContractFunctionWithValueNoEncoding(
+						test.call().signature, test.call().value.value, test.call().arguments.rawBytes());
```
`);` on a line of its own.

chriseth commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -88,156 +103,180 @@ SemanticTest::SemanticTest(string const& _filename, langutil::EVMVersion _evmVer
-			throw;
+			m_compileViaYulCanBeSet = true;
+			AnsiColorized(_stream, _formatted, {BOLD, YELLOW})
+				<< _linePrefix << endl
```
`<<` on the previous line.

chriseth commented on test/tools/isoltest.cpp:
```diff
@@ -424,14 +425,27 @@ int main(int argc, char const *argv[])
-		cout << solidity::test::evmoneDownloadLink << endl;
+		disableSemantics = !solidity::test::EVMHost::checkVmPaths(options.vmPaths);
+	}
+	catch (std::exception const& _exception)
```
What exception can be throw there?

aarlt replied:
`std::runtime_error`, so I will only check for that one.
